text/0000-hidden-impls.md

```diff
@@ -0,0 +1,878 @@
+# Draft RFC: Hidden implementations
+
+- Feature Name: `hidden_impl`
+- Start Date: 2018-06-24
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Allow a visibility modifier on a `trait` implementation such that:
+
++ it is public wrt. coherence and overlap
++ it is not usable in contexts where it is *hidden* (not visible)
+
+An example:
+
+```rust
+#[derive(Clone)]
+struct Foo(..);
+
+crate impl Copy for Foo {} // Can only be used in the current crate!
+```
+
+# Motivation
+[motivation]: #motivation
+
+First, let's assume that we have a trait `Property` defined somewhere.
+This trait, which doesn't have to be inside your crate, is also visible
+outside of it. It also happens that your crate has defined and exported
+a type called `Thing`. The specifics of `Property` and `Thing` are not
+interesting for our purposes, but they exist.
+
+You now want to, define an implementation of `Property` for `Thing`.
+But for some reason, you are not willing, or able, to expose this implementation
+outside your crate and make it part of your semantic versioning guarantees.
+
+An example of such a scenario is defining a new-type wrapper `TestRng` around
+`XorShiftRng` of the `rand` crate. We want to define an implementation of
+`SeedableRng` for `TestRng` to take advantage of the logic of `FromEntropy`
+which is automatically provided for us when `Rng: SeedableRng`.
+But we also want to avoid exposing `SeedableRng` as a guarantee to the
+users of our library (`proptest`) right now. The allure of getting the logic
+provided when defining `SeedableRng` could also have caused us to prematurely
+and mistakenly provide a guarantee we were not ready to provide.
+
+Another example use-case which [@nikomatsakis](https://github.com/nikomatsakis)
+has provided is wanting to have the convenience of `#[derive(Default)]`,
+but not wanting to commit to there being a default value.
+
+However, we have no way eat our cake and keep it. We can't segregate who
+has access to what implementations by visibility.
```

Comment:
One thing to potentially add/explore perhaps is to take a look at the compiler's `TyCtxt` and similar types that are used throughout, but can't currently be implemented directly on outside of `librustc`; it feels loosely like that's another significant motivation -- currently there's lots of functions that should/could be inherent on tcx but can't be.